File: src/main/java/frc/robot/commands/DefaultDriveCmd.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.commands;

import static edu.wpi.first.units.Units.MetersPerSecond;
import static edu.wpi.first.units.Units.RadiansPerSecond;
import static edu.wpi.first.units.Units.RotationsPerSecond;

import com.ctre.phoenix6.swerve.SwerveModule.DriveRequestType;
import com.ctre.phoenix6.swerve.SwerveRequest;
import edu.wpi.first.math.MathUtil;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.DriverStation.Alliance;
import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj2.command.Command;
import frc.robot.generated.TunerConstants;
import frc.robot.subsystems.CommandSwerveDrivetrain;
import frc.robot.util.constants.SwerveConstants;
import java.util.Optional;
import java.util.function.BooleanSupplier;
import java.util.function.Consumer;
import java.util.function.DoubleConsumer;
import java.util.function.DoubleSupplier;
import java.util.function.Supplier;

/* You should consider using the more terse Command factories API instead https://docs.wpilib.org/en/stable/docs/software/commandbased/organizing-command-based.html#defining-commands */
public class DefaultDriveCmd extends Command {

  private static class Speeds {
    double translation;
    double strafe;
    double rotation;
  }

  private final CommandSwerveDrivetrain swerve;
  private final DoubleSupplier translationSup;
  private final DoubleSupplier strafeSup;
  private final DoubleSupplier rotationSup;
  private final BooleanSupplier halfSpeedSup;

  private final Supplier<Optional<Rotation2d>> headingGetter;
  private final Consumer<Optional<Rotation2d>> headingSetter;
  private final DoubleSupplier rotationLastTriggeredGetter;
  private final DoubleConsumer rotationLastTriggeredSetter;

  private double maxSpeed;
  private double maxAngularRate;

  private final SwerveRequest.FieldCentric drive;
  private final SwerveRequest.FieldCentricFacingAngle driveMaintainHeading;

  /** Creates a new DefaultDriveCmd. */
  public DefaultDriveCmd(
      CommandSwerveDrivetrain swerve,
      DoubleSupplier translationSup,
      DoubleSupplier strafeSup,
      DoubleSupplier rotationSup,
      BooleanSupplier halfSpeedSup,
      Supplier<Optional<Rotation2d>> headingGetter,
      Consumer<Optional<Rotation2d>> headingSetter,
      DoubleSupplier rotationLastTriggeredGetter,
      DoubleConsumer rotationLastTriggeredSetter) {

    this.swerve = swerve;
    this.translationSup = translationSup;
    this.strafeSup = strafeSup;
    this.rotationSup = rotationSup;
    this.halfSpeedSup = halfSpeedSup;
    this.headingGetter = headingGetter;
    this.headingSetter = headingSetter;
    this.rotationLastTriggeredGetter = rotationLastTriggeredGetter;
    this.rotationLastTriggeredSetter = rotationLastTriggeredSetter;

    maxSpeed = TunerConstants.kSpeedAt12Volts.in(MetersPerSecond); // Max speed at 12 volts
    maxAngularRate =
        RotationsPerSecond.of(1).in(RadiansPerSecond); // 1 rotation per second max angular velocity

    drive =
        new SwerveRequest.FieldCentric()
            .withDeadband(maxSpeed * 0.05)
            .withRotationalDeadband(maxAngularRate * 0.05)
            .withDriveRequestType(DriveRequestType.OpenLoopVoltage)
            .withDesaturateWheelSpeeds(true);

    driveMaintainHeading =
        new SwerveRequest.FieldCentricFacingAngle()
            .withDeadband(maxSpeed * 0.05)
            .withRotationalDeadband(maxAngularRate * 0.05)
            .withDriveRequestType(DriveRequestType.OpenLoopVoltage)
            .withDesaturateWheelSpeeds(true);

    driveMaintainHeading.HeadingController.setPID(
        SwerveConstants.HEADING_KP, SwerveConstants.HEADING_KI, SwerveConstants.HEADING_KD);
    driveMaintainHeading.HeadingController.enableContinuousInput(-Math.PI, Math.PI);
    driveMaintainHeading.HeadingController.setTolerance(SwerveConstants.HEADING_TOLERANCE);

    addRequirements(swerve);
  }

  // Called when the command is initially scheduled.
  @Override
  public void initialize() {}

  // Called every time the scheduler runs while the command is scheduled.
  @Override
  public void execute() {
    double rawRotation = rotationSup.getAsDouble();
    var speeds =
        processJoystickInputs(
            translationSup.getAsDouble(),
            strafeSup.getAsDouble(),
            rawRotation,
            halfSpeedSup.getAsBoolean());

    boolean rotationTriggered = Math.abs(rawRotation) > SwerveConstants.SWERVE_DEADBAND;
    boolean rotationActive =
        MathUtil.isNear(rotationLastTriggeredGetter.getAsDouble(), Timer.getFPGATimestamp(), 0.1)
            && (Math.abs(swerve.getState().Speeds.omegaRadiansPerSecond) > Math.toRadians(10));

    if (rotationTriggered) {
      rotationLastTriggeredSetter.accept(Timer.getFPGATimestamp());
    }

    if (rotationTriggered || rotationActive) {
      setSwerveToRotate(speeds.translation, speeds.strafe, speeds.rotation);
    } else {
      setSwerveToMaintainHeading(speeds.translation, speeds.strafe);
    }
  }

  // Called once the command ends or is interrupted.
  @Override
  public void end(boolean interrupted) {}

  // Returns true when the command should end.
  @Override
  public boolean isFinished() {
    return false;
  }

  private Speeds processJoystickInputs(
      double rawTranslation, double rawStrafe, double rawRotation, boolean halfSpeed) {
    double translation = MathUtil.applyDeadband(rawTranslation, SwerveConstants.SWERVE_DEADBAND, 1);
    double strafe = MathUtil.applyDeadband(rawStrafe, SwerveConstants.SWERVE_DEADBAND, 1);
    double rotation = MathUtil.applyDeadband(rawRotation, SwerveConstants.SWERVE_DEADBAND, 1);

    translation = Math.copySign(translation * translation, translation);
    strafe = Math.copySign(strafe * strafe, strafe);
    rotation = Math.copySign(rotation * rotation, rotation);

    if (halfSpeed) {
      translation *= 0.35;
      strafe *= 0.35;
      rotation *= 0.35;
    }

    Speeds speeds = new Speeds();
    speeds.translation = translation * maxSpeed;
    speeds.strafe = strafe * maxSpeed;
    speeds.rotation = rotation * maxAngularRate;
    return speeds;
  }

  private void setSwerveToRotate(double translation, double strafe, double rotation) {
    swerve.setControl(
        drive.withVelocityX(translation).withVelocityY(strafe).withRotationalRate(rotation));
    headingSetter.accept(Optional.empty()); // was: currentHeading = Optional.empty();
  }

  private void setSwerveToMaintainHeading(double translation, double strafe) {
    if (headingGetter.get().isEmpty()) { // was: currentHeading.isEmpty()
      headingSetter.accept(Optional.of(swerve.getState().Pose.getRotation()));
    }

    Optional<Alliance> alliance = DriverStation.getAlliance();
    if (alliance.isPresent()) {
      Rotation2d targetDirection =
          alliance.get() == Alliance.Blue
              ? headingGetter.get().get() // was: currentHeading.get()
              : headingGetter.get().get().rotateBy(Rotation2d.fromDegrees(180));
      swerve.setControl(
          driveMaintainHeading
              .withVelocityX(translation)
              .withVelocityY(strafe)
              .withTargetDirection(targetDirection));
    }
  }
}

File: src/main/cpp/commands/DefaultDriveCmd.cpp
#include "commands/DefaultDriveCmd.h"

#include <numbers>

#include <frc/DriverStation.h>
#include <frc/MathUtil.h>
#include <frc/Timer.h>

#include "Constants.h"
#include "generated/TunerConstants.h"
#include "subsystems/CommandSwerveDrivetrain.h"

using namespace ctre::phoenix6;

DefaultDriveCmd::DefaultDriveCmd(
    CommandSwerveDrivetrain* swerve,
    std::function<double()> translationSup,
    std::function<double()> strafeSup,
    std::function<double()> rotationSup,
    std::function<bool()> halfSpeedSup,
    std::function<std::optional<frc::Rotation2d>()> headingGetter,
    std::function<void(std::optional<frc::Rotation2d>)> headingSetter,
    std::function<double()> rotationLastTriggeredGetter,
    std::function<void(double)> rotationLastTriggeredSetter)
    : m_swerve{swerve},
      m_translationSup{std::move(translationSup)},
      m_strafeSup{std::move(strafeSup)},
      m_rotationSup{std::move(rotationSup)},
      m_halfSpeedSup{std::move(halfSpeedSup)},
      m_headingGetter{std::move(headingGetter)},
      m_headingSetter{std::move(headingSetter)},
      m_rotationLastTriggeredGetter{std::move(rotationLastTriggeredGetter)},
      m_rotationLastTriggeredSetter{std::move(rotationLastTriggeredSetter)},
      m_maxSpeed{TunerConstants::kSpeedAt12Volts.value()},
      m_maxAngularRate{2.0 * std::numbers::pi}  // 1 rotation/s in rad/s
{
  m_drive.Deadband = m_maxSpeed * 0.05;
  m_drive.RotationalDeadband = m_maxAngularRate * 0.05;
  m_drive.DriveRequestType =
      swerve::SwerveModule::DriveRequestType::OpenLoopVoltage;
  m_drive.DesaturateWheelSpeeds = true;

  m_driveMaintainHeading.Deadband = m_maxSpeed * 0.05;
  m_driveMaintainHeading.RotationalDeadband = m_maxAngularRate * 0.05;
  m_driveMaintainHeading.DriveRequestType =
      swerve::SwerveModule::DriveRequestType::OpenLoopVoltage;
  m_driveMaintainHeading.DesaturateWheelSpeeds = true;

  m_driveMaintainHeading.HeadingController.SetPID(
      SwerveConstants::kHeadingKP, SwerveConstants::kHeadingKI,
      SwerveConstants::kHeadingKD);
  m_driveMaintainHeading.HeadingController.EnableContinuousInput(-std::numbers::pi,
                                                                  std::numbers::pi);
  m_driveMaintainHeading.HeadingController.SetTolerance(
      SwerveConstants::kHeadingTolerance);

  AddRequirements(swerve);
}

void DefaultDriveCmd::Initialize() {}

void DefaultDriveCmd::Execute() {
  double rawRotation = m_rotationSup();
  auto speeds = ProcessJoystickInputs(m_translationSup(), m_strafeSup(),
                                      rawRotation, m_halfSpeedSup());

  bool rotationTriggered =
      std::abs(rawRotation) > SwerveConstants::kSwerveDeadband;
  bool rotationActive =
      std::abs(m_rotationLastTriggeredGetter() - frc::Timer::GetFPGATimestamp().value()) < 0.1 &&
      std::abs(m_swerve->GetState().Speeds.omega.value()) > (10.0 * std::numbers::pi / 180.0);

  if (rotationTriggered) {
    m_rotationLastTriggeredSetter(frc::Timer::GetFPGATimestamp().value());
  }

  if (rotationTriggered || rotationActive) {
    SetSwerveToRotate(speeds.translation, speeds.strafe, speeds.rotation);
  } else {
    SetSwerveToMaintainHeading(speeds.translation, speeds.strafe);
  }
}

void DefaultDriveCmd::End(bool interrupted) {}

bool DefaultDriveCmd::IsFinished() {
  return false;
}

DefaultDriveCmd::Speeds DefaultDriveCmd::ProcessJoystickInputs(
    double rawTranslation, double rawStrafe, double rawRotation, bool halfSpeed) {
  double translation =
      frc::ApplyDeadband(rawTranslation, SwerveConstants::kSwerveDeadband);
  double strafe =
      frc::ApplyDeadband(rawStrafe, SwerveConstants::kSwerveDeadband);
  double rotation =
      frc::ApplyDeadband(rawRotation, SwerveConstants::kSwerveDeadband);

  translation = std::copysign(translation * translation, translation);
  strafe = std::copysign(strafe * strafe, strafe);
  rotation = std::copysign(rotation * rotation, rotation);

  if (halfSpeed) {
    translation *= 0.35;
    strafe *= 0.35;
    rotation *= 0.35;
  }

  return Speeds{translation * m_maxSpeed, strafe * m_maxSpeed,
                rotation * m_maxAngularRate};
}

void DefaultDriveCmd::SetSwerveToRotate(double translation, double strafe,
                                        double rotation) {
  m_swerve->SetControl(m_drive.WithVelocityX(units::meters_per_second_t{translation})
                           .WithVelocityY(units::meters_per_second_t{strafe})
                           .WithRotationalRate(units::radians_per_second_t{rotation}));
  m_headingSetter(std::nullopt);
}

void DefaultDriveCmd::SetSwerveToMaintainHeading(double translation,
                                                  double strafe) {
  if (!m_headingGetter().has_value()) {
    m_headingSetter(m_swerve->GetState().Pose.Rotation());
  }

  auto alliance = frc::DriverStation::GetAlliance();
  auto heading = m_headingGetter();
  if (alliance.has_value() && heading.has_value()) {
    frc::Rotation2d targetDirection =
        alliance.value() == frc::DriverStation::Alliance::kBlue
            ? heading.value()
            : heading.value().RotateBy(frc::Rotation2d{units::degree_t{180}});
    m_swerve->SetControl(
        m_driveMaintainHeading.WithVelocityX(units::meters_per_second_t{translation})
            .WithVelocityY(units::meters_per_second_t{strafe})
            .WithTargetDirection(targetDirection));
  }
}
